cql3/expr/expression.hh

```diff
@@ -183,14 +191,53 @@ struct function_call {
 
 struct cast {
     nested_expression arg;
-    cql3_type type;
+    std::variant<cql3_type, shared_ptr<cql3_type::raw>> type;
 };
 
 struct field_selection {
     nested_expression structure;
     shared_ptr<column_identifier_raw> field;
 };
 
+struct null {
+};
+
+struct bind_variable {
+    struct scalar {};
+    struct scalar_in {};
+    struct tuple {};
+    struct tuple_in {};
+    // FIXME: infer shape from expression rather than from grammar
+    std::variant<scalar, scalar_in, tuple, tuple_in> shape;
+    int32_t bind_index;
+};
+
+// A constant which does not yet have a date type. It is partially typed
+// (we know if it's floating or int) but not sized.
+struct untyped_constant {
+    enum type_class { integer, floating_point, string, boolean, duration, uuid, hex };
```

Comment:
Nitpick: this would be nicer as a super_enum of cql3_type::kind.

Reply:
No, cql3_type::kind describes types, and this describes type classes (e.g. int/smallint/bigint/varint are all integers).

Reply:
Doesn't super_enum definition allow using just a subset?

Reply:
Maybe, but `integer` here doesn't map to any cql3_type::kind member. It maps to all of the integer ones.

In any case, I'd like to disappear it an replace it by a typed constant, but this series tries not to do too many changes at once. Replacing it with a typed constant will require a widening or narrowing pass later (depending on whether we represent constants in the narrowest or widest possible form).

Reply:
OK.